pulsar-broker/src/test/java/org/apache/pulsar/client/api/DeduplicationEndToEndTest.java

```diff
@@ -0,0 +1,672 @@
+        // this new message will be routed to the same partition, so the message will not be duplicated.
+        producer.newMessage().value(data).sequenceId(lastId).key(String.valueOf(lastId)).sendAsync().exceptionally(e -> {
+            // should not enter here
+            Assert.fail();
+            return null;
+        }).get();
+
+        // consume the message, there are only one messages in the topic
+        Message<byte[]> message = assertNotDuplicate(consumer, data);
+        assertEquals(message.getSequenceId(), lastId);
+
+        // clean up
+        producer.close();
+        consumer.close();
+    }
+
+    /**
+     * trigger the partition number update for the producer
+     * @param producer
+     * @param expectedPartitionCount
+     * @throws Exception
+     */
+    private void triggerPartitionUpdateForPartitionedProducer(PartitionedProducerImpl<byte[]> producer, int expectedPartitionCount) throws Exception {
+        Field partitionsAutoUpdateTimerTask = PartitionedProducerImpl.class.getDeclaredField("partitionsAutoUpdateTimerTask");
+        partitionsAutoUpdateTimerTask.setAccessible(true);
+        TimerTask timerTask = (TimerTask) partitionsAutoUpdateTimerTask.get(producer);
+        ((PulsarClientImpl) pulsarClient).getTimer().newTimeout(timerTask, 0, TimeUnit.MILLISECONDS);
+        Awaitility.await().until(() -> {
+            try {
+                return producer.getNumOfPartitions() == expectedPartitionCount;
+            } catch (Exception e) {
+                return false;
+            }
+        });
+    }
+
+    private void triggerPartitionUpdateForPartitionedConsumer(MultiTopicsConsumerImpl<byte[]> consumer, int expectedPartitionCount) throws Exception {
+        Field partitionsAutoUpdateTimerTask = MultiTopicsConsumerImpl.class.getDeclaredField("partitionsAutoUpdateTimerTask");
+        partitionsAutoUpdateTimerTask.setAccessible(true);
+        TimerTask timerTask = (TimerTask) partitionsAutoUpdateTimerTask.get(consumer);
+        ((PulsarClientImpl) pulsarClient).getTimer().newTimeout(timerTask, 0, TimeUnit.MILLISECONDS);
+        Awaitility.await().until(() -> {
+            try {
+                return consumer.getPartitions().size() == expectedPartitionCount;
+            } catch (Exception e) {
+                return false;
+            }
+        });
+    }
+
+    /**
+     * simulate the case where the producer sends the message but doesn't receive the ack
+     * due to network issue, broker issue, etc. User receives the exception and sends the same message again.
+     * With deduplication enabled and user control sequence id, the topic is multi partitioned,
+     * though we use the key based routing to route the messages with the same key to the same partition,
+     * If the partition number is updated between the two messages, the two messages will be routed to different partitions,
+     * so the message is duplicated.
```

Comment:
I don't recommend adding such a test that verifies a feature does not work for certain cases.

To make users know message deduplication only works for a single partition, the best way is to just add a documentation for that.

Reply:
We need a way to elaborate various cases where message dudup can work well and not.
I think we need to update the document.
Before i research into the implementation, i read the official document and the `Limitation of idempotent producer` is described as:
https://pulsar.apache.org/docs/next/txn-why/
```
Due to the monotonic increasing sequence ID, this solution only works on a single partition and within a single producer session (that is, for producing one message), so there is no atomicity when producing multiple messages to one or multiple partitions.

In this case, if there are some failures (for example, client / broker / bookie crashes, network failure, and more) in the process of producing and receiving messages, messages are re-processed and re-delivered, which may cause data loss or data duplication:

For the producer: if the producer retries sending messages, some messages are persisted multiple times; if the producer does not retry sending messages, some messages are persisted once and other messages are lost.

For the consumer: since the consumer does not know whether the broker has received messages or not, the consumer may not retry sending acks, which causes it to receive duplicate messages.
```
I can't get anything clear about what problem it can't fix. **Text description can not describe the situations precisely**. We need such kind of test to prove out the ability and the shortcoming of features.

Reply:
> Text description can not describe the situations precisely. We need such kind of test to prove out the ability and the shortcoming of features.

Reading the tests from source code is never the first way to know a feature. The main effect of tests is to prevent breaking changes. In short, we should
- Improve the documents if the feature is unclear to users, including the limitation
- Add the tests if what the feature guarantees is not protected by tests

Reply:
No matter how, we need a way to **present these test code to users** to show what cases it can solve and not. For example, though i have summarized the cases in the comments of the test, developers still need to read the concrete test code to be clear about what this case really is. 
Even many professional developers of pulsar are clear about the limitation of this feature only after reading these tests.

Reply:
The most proper approach might be contributing to the https://github.com/apache/pulsar-site repo